File: String/Atoi.java
public class Atoi {
    public int myAtoi(String s) {
        int idx = 0;
        int n = s.length();
        int result = 0;
        int sign = 1;

        // Ignore leading whitespace
        while (idx < n && s.charAt(idx) == ' ') {
            idx++;
        }

        if (idx < n && (s.charAt(idx) == '+' || s.charAt(idx) == '-')) {
            sign = (s.charAt(idx) == '-') ? -1 : 1;
            idx++;
        }

        while (idx < n && Character.isDigit(s.charAt(idx))) {
            int digit = s.charAt(idx) - '0';

            if (result > (Integer.MAX_VALUE - digit) / 10) {
                return (sign == 1) ? Integer.MAX_VALUE : Integer.MIN_VALUE;
            }

            result = result * 10 + digit;
            idx++;
        }
        return result * sign;
    }

    public static void main(String[] args) {
        Atoi solution = new Atoi();

        // Test cases
        System.out.println(solution.myAtoi("42")); // 42
        System.out.println(solution.myAtoi("   -042")); // -42
        System.out.println(solution.myAtoi("1337c0d3")); // 1337
        System.out.println(solution.myAtoi("0-1")); // 0
        System.out.println(solution.myAtoi("words and 987")); // 0
    }
}

File: String/Atoi.cpp
#include <iostream>
#include <string>
#include <climits> // for INT_MAX and INT_MIN
using namespace std;

class Atoi {
public:
    int myAtoi(const string& s) {
        int idx = 0;
        int n = s.length();
        int result = 0;
        int sign = 1;

        // Ignore leading whitespace
        while (idx < n && s[idx] == ' ') {
            idx++;
        }

        if (idx < n && (s[idx] == '+' || s[idx] == '-')) {
            sign = (s[idx] == '-') ? -1 : 1;
            idx++;
        }

        while (idx < n && isdigit(s[idx])) {
            int digit = s[idx] - '0';

            if (result > (INT_MAX - digit) / 10) {
                return (sign == 1) ? INT_MAX : INT_MIN;
            }

            result = result * 10 + digit;
            idx++;
        }

        return result * sign;
    }
};

int main() {
    Atoi solution;

    // Test cases
    cout << solution.myAtoi("42") << endl;         // 42
    cout << solution.myAtoi("   -042") << endl;    // -42
    cout << solution.myAtoi("1337c0d3") << endl;   // 1337
    cout << solution.myAtoi("0-1") << endl;        // 0
    cout << solution.myAtoi("words and 987") << endl; // 0

    return 0;
}
